plugins/cache/cache_main.ml

```diff
@@ -6,246 +6,120 @@ include Self()
-        match diff with
-        | `Right _ -> ()
-        | `Left e | `Unequal (e,_) -> remove_entry e)
-
-  module T = struct
-    type t = index [@@deriving bin_io]
-  end
-
-  let from_file : type t.
-    (module Binable.S with type t = t) -> string -> t = fun b file ->
-    let module T = (val b) in
-    let fd = Unix.(openfile file [O_RDONLY] 0o400) in
-    try
-      let data = Mmap.V1.map_file
-          fd Bigarray.char Bigarray.c_layout false [|-1|] in
-      let pos_ref = ref 0 in
-      let t = T.bin_read_t (Bigarray.array1_of_genarray data) ~pos_ref in
-      Unix.close fd;
-      t
-    with e -> Unix.close fd; raise e
-  [@@warning "-D"]
-
-  let open_temp () =
-    let tmp =
-      Filename.temp_file ~temp_dir:(cache_dir ()) "tmp" "index" in
-    try tmp, Unix.(openfile tmp [O_RDWR] 0o600)
-    with e -> Sys.remove tmp; raise e
 
-  let to_file : type t.
-    (module Binable.S with type t = t) -> string -> t -> unit =
-    fun b file data ->
-      let module T = (val b) in
-      let tmp,fd = open_temp () in
-      let size = T.bin_size_t data in
-      let () =
-        try
-          let buf =
-            Mmap.V1.map_file
-              fd Bigarray.char Bigarray.c_layout true [|size|] in
-          let _ = T.bin_write_t (Bigarray.array1_of_genarray buf) ~pos:0 data in
-          Unix.close fd
-        with e -> Unix.close fd; Sys.remove tmp; raise e in
-      Sys.rename tmp file
-  [@@warning "-D"]
-
-  let index_of_file file =
-    try from_file (module T) file
-    with e ->
-      warning "read index: %s" (Exn.to_string e);
-      empty
-
-  let index_to_file file index =
-    try to_file (module T) file index
-    with e -> warning "store index: %s" (Exn.to_string e)
-
-  let with_index ~f =
-    let cache_dir = cache_dir () in
-    let file = cache_dir / index_file in
-    let lock = cache_dir / lock_file in
-    let lock = Unix.openfile lock Unix.[O_RDWR; O_CREAT] 0o640 in
-    Unix.lockf lock Unix.F_LOCK 0;
-    protect ~f:(fun () ->
-        let init = index_of_file file in
-        let index',data = f cache_dir init in
-        remove_files init index';
-        let index = clean index' in
-        remove_files index' index;
-        index_to_file file index;
+  let t = ref None
+
+  let mtime () =
+    let index = Index.index_file () in
+    if Sys.file_exists index then
+      let s = Unix.stat index in
+      Some Unix.(s.st_mtime)
+    else None
+
+  let update index =
+    let mtime = mtime () in
+    t := Some {index; mtime }
+
+  let read () = match !t with
+    | None ->
+      let idx = Index.read () in
+      let idx =
+        if idx.config.gc_enabled then
+          GC.run idx
+        else idx in
+      update idx;
+      idx
+    | Some {index} -> index
+
+  let force_read () =
+    t := None;
+    read ()
+
+  let write index =
+    Index.write index;
+    update index
+
+  let reload () =
+    match !t with
+    | None -> read ()
+    | Some t ->
+      if t.mtime = mtime () then t.index
```

Comment:
don't compare floats for equality, use `<` or `>` or `<=`, etc. And don't use polymorphic compare with floats.